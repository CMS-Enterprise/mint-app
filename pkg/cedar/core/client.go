package cedarcore

import (
	"context"
	"net/http"
	"time"

	ld "gopkg.in/launchdarkly/go-server-sdk.v5"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	cache "github.com/patrickmn/go-cache"
	"go.uber.org/zap"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/core/gen/client"
	"github.com/cmsgov/easi-app/pkg/flags"
)

const (
	cedarCoreEnabledKey     = "cedarCoreEnabled"
	cedarCoreEnabledDefault = false
)

// NewClient builds the type that holds a connection to the CEDAR Core API
func NewClient(ctx context.Context, cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
	fnEmit := func(ctx context.Context) bool {
		lduser := flags.Principal(ctx)
		result, err := ldClient.BoolVariation(cedarCoreEnabledKey, lduser, cedarCoreEnabledDefault)
		if err != nil {
			appcontext.ZLogger(ctx).Info(
				"problem evaluating feature flag",
				zap.Error(err),
				zap.String("flagName", cedarCoreEnabledKey),
				zap.Bool("flagDefault", cedarCoreEnabledDefault),
				zap.Bool("flagResult", result),
			)
		}
		return result
	}

	c := cache.New(cache.NoExpiration, cache.NoExpiration) // Don't expire data _or_ clean it up

	hc := http.DefaultClient

	client := &Client{
		cedarCoreEnabled: fnEmit,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc:    hc,
		cache: c,
	}

	// Start cache refresh for systems
	client.startCacheRefresh(ctx, time.Minute*5, client.populateSystemSummaryCache)

	return client
}

// Client represents a connection to the CEDAR Core API
type Client struct {
	cedarCoreEnabled func(context.Context) bool
	auth             runtime.ClientAuthInfoWriter
	sdk              *apiclient.CEDARCoreAPI
	hc               *http.Client
	cache            *cache.Cache
}

// startCacheRefresh starts a goroutine that will run `populateCache` based on cacheRefreshTime
// This function returns no errors, and only logs when something goes wrong
func (c *Client) startCacheRefresh(ctx context.Context, cacheRefreshTime time.Duration, populateCache func(context.Context) error) {
	ticker := time.NewTicker(cacheRefreshTime)
	go func(ctx context.Context) {
		for {
			err := populateCache(ctx)
			if err != nil {
				appcontext.ZLogger(ctx).Error("Failed to refresh cache", zap.Error(err))
			}
			// Wait for the ticker. This will block the current goroutine until the ticker sends a message over the channel
			<-ticker.C
		}
	}(ctx)
}
