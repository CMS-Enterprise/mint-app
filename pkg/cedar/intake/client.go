// Package intake serves as an Anti-Corruption Layer (in Domain Driven Design
// parlance), keeping the implementation details of working with the
// downstream CEDAR Intake API well separated, so the rest of the internal
// workings of the EASi code are not using anything defined by the
// external API
package intake

import (
	"context"
	"fmt"
	"net/http"
	"strconv"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	"go.uber.org/zap"
	ld "gopkg.in/launchdarkly/go-server-sdk.v5"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client/health_check"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client/intake"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/translation"
	"github.com/cmsgov/easi-app/pkg/flags"
	"github.com/cmsgov/easi-app/pkg/models"
)

const (
	emitFlagKey = "emit-to-cedar"
	emitDefault = false
)

// NewClient builds the type that holds a connection to the CEDAR Intake API
func NewClient(cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
	fnEmit := func(ctx context.Context) bool {
		// this is the conditional way of stopping us from submitting to CEDAR; see EASI-1025
		lduser := flags.Principal(ctx)
		result, err := ldClient.BoolVariation(emitFlagKey, lduser, emitDefault)
		if err != nil {
			appcontext.ZLogger(ctx).Info(
				"problem evaluating feature flag",
				zap.Error(err),
				zap.String("flagName", emitFlagKey),
				zap.Bool("flagDefault", emitDefault),
				zap.Bool("flagResult", result),
			)
		}
		return result
	}

	hc := http.DefaultClient

	return &Client{
		emitToCedar: fnEmit,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc: hc,
	}
}

// Client represents a connection to the CEDAR Intake API
type Client struct {
	emitToCedar func(context.Context) bool
	auth        runtime.ClientAuthInfoWriter
	sdk         *apiclient.CEDARIntake
	hc          *http.Client
}

// CheckConnection hits the CEDAR Intake API `/healthcheck` endpoint to verify
// that our connection is functional
func (c *Client) CheckConnection(ctx context.Context) error {
	if !c.emitToCedar(ctx) {
		return nil
	}

	params := health_check.NewHealthCheckParamsWithContext(ctx)
	params.HTTPClient = c.hc

	resp, err := c.sdk.HealthCheck.HealthCheck(params, c.auth)
	if err != nil {
		return err
	}
	if resp.Payload == nil {
		return fmt.Errorf("no response body received")
	}
	return nil
}

// PublishSystemIntake sends a system intake to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishSystemIntake(ctx context.Context, si models.SystemIntake) error {
	intakeObject := translation.TranslatableSystemIntake(si)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishBusinessCase sends a business case to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishBusinessCase(ctx context.Context, bc models.BusinessCase) error {
	intakeObject := translation.TranslatableBusinessCase(bc)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishGRTFeedback sends an item of GRT feedback to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishGRTFeedback(ctx context.Context, feedback models.GRTFeedback) error {
	intakeObject := translation.TranslatableFeedback(feedback)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishAction sends an action to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishAction(ctx context.Context, action models.Action) error {
	intakeObject := translation.TranslatableAction(action)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishNote sends a note to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishNote(ctx context.Context, note models.Note) error {
	intakeObject := translation.TranslatableNote(note)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// private method for publishing anything that satisfies the translation.IntakeObject interface to CEDAR through the Intake API
func (c *Client) publishIntakeObject(ctx context.Context, model translation.IntakeObject) error {
	// constant values for now; may become non-constant if/when we revisit handling CEDAR validation errors

	// from the Swagger (cedar_intake.json, definitions/IntakeInput/properties/version):
	// "The version associated with the object in the body. This value can be incremented in the event a transaction needs to be resubmitted."
	const objectVersion = 1

	// from the Swagger (cedar_intake.json, paths/"/intake"/post/parameters/validatePayload):
	// "Determines if schema validation of the payload is performed syncronously before persisting the record or asyncronously after the record has been persisted"
	const isValidatedSynchronously = false // false for V1 schemas; this may change when we implement V2 schemas for EASI-1614

	id := model.ObjectID()
	objectType := model.ObjectType()

	if !c.emitToCedar(ctx) {
		appcontext.ZLogger(ctx).Info("snapshot publishing disabled", zap.String("object ID", id), zap.String("object type", objectType))
		return nil
	}

	input, err := model.CreateIntakeModel()

	if err != nil {
		appcontext.ZLogger(ctx).Error(
			"problem building cedar payload",
			zap.String("object ID", id),
			zap.String("object type", objectType),
			zap.Error(err),
		)
		return nil
	}

	params := intake.NewIntakeAddParamsWithContext(ctx)
	params.HTTPClient = c.hc
	params.Body = input

	params.ValidatePayload = strconv.FormatBool(isValidatedSynchronously)

	objectVersionStr := strconv.FormatInt(objectVersion, 10)
	params.Body.Version = &objectVersionStr

	resp, err := c.sdk.Intake.IntakeAdd(params, c.auth)
	if err != nil {
		return err
	}

	// TODO: we may need to read through the results for any "duplicate entry" error codes
	// that CEDAR _may_ send us, especially if we are re-transmitting entities... We
	// don't really care if there are dupes on their side, because we do NOT want to hold
	// any state on our side about who/what/when may have transmitted a previous copy
	// of the info to CEDAR.
	// appcontext.ZLogger(ctx).Info(fmt.Sprintf("results: %v", resp.Payload))
	_ = resp

	return nil
}
