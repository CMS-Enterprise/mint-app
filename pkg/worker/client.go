package worker

import (
	"fmt"

	faktory "github.com/contribsys/faktory/client"
)

// NewClient returns a new faktory Client.
func NewClient() (*faktory.Client, error) {
	cl, err := faktory.Open()
	if err != nil {
		return nil, fmt.Errorf("error opening faktory client: %w", err)
	}
	return cl, nil
}
