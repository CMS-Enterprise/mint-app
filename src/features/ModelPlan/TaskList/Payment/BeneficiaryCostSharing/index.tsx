import React, { useRef } from 'react';
import { useTranslation } from 'react-i18next';
import { useHistory, useParams } from 'react-router-dom';
import {
  Button,
  Fieldset,
  Grid,
  GridContainer,
  Icon,
  Label
} from '@trussworks/react-uswds';
import { NotFoundPartial } from 'features/NotFound';
import { Field, Form, Formik, FormikProps } from 'formik';
import {
  ClaimsBasedPayType,
  GetBeneficiaryCostSharingQuery,
  PayType,
  TypedUpdatePaymentsDocument,
  useGetBeneficiaryCostSharingQuery
} from 'gql/generated/graphql';

import AddNote from 'components/AddNote';
import AskAQuestion from 'components/AskAQuestion';
import BooleanRadio from 'components/BooleanRadioForm';
import Breadcrumbs, { BreadcrumbItemOptions } from 'components/Breadcrumbs';
import ConfirmLeave from 'components/ConfirmLeave';
import FieldGroup from 'components/FieldGroup';
import MutationErrorModal from 'components/MutationErrorModal';
import PageHeading from 'components/PageHeading';
import PageNumber from 'components/PageNumber';
import TextAreaField from 'components/TextAreaField';
import useHandleMutation from 'hooks/useHandleMutation';
import usePlanTranslation from 'hooks/usePlanTranslation';

import { renderCurrentPage, renderTotalPages } from '..';

type BeneficiaryCostSharingFormType =
  GetBeneficiaryCostSharingQuery['modelPlan']['payments'];

const BeneficiaryCostSharing = () => {
  const { t: paymentsT } = useTranslation('payments');

  const { t: paymentsMiscT } = useTranslation('paymentsMisc');

  const { t: miscellaneousT } = useTranslation('miscellaneous');

  const {
    waiveBeneficiaryCostSharingForAnyServices:
      waiveBeneficiaryCostSharingForAnyServicesConfig,
    waiverOnlyAppliesPartOfPayment: waiverOnlyAppliesPartOfPaymentConfig
  } = usePlanTranslation('payments');

  const { modelID } = useParams<{ modelID: string }>();

  const formikRef = useRef<FormikProps<BeneficiaryCostSharingFormType>>(null);
  const history = useHistory();

  const { data, loading, error } = useGetBeneficiaryCostSharingQuery({
    variables: {
      id: modelID
    }
  });

  const {
    id,
    payType,
    payClaims,
    beneficiaryCostSharingLevelAndHandling,
    waiveBeneficiaryCostSharingForAnyServices,
    waiveBeneficiaryCostSharingServiceSpecification,
    waiverOnlyAppliesPartOfPayment,
    waiveBeneficiaryCostSharingNote
  } = (data?.modelPlan?.payments || {}) as BeneficiaryCostSharingFormType;

  const modelName = data?.modelPlan?.modelName || '';

  const { mutationError } = useHandleMutation(TypedUpdatePaymentsDocument, {
    id,
    formikRef
  });

  const nextPage = () => {
    if (
      formikRef?.current?.values.payType.includes(
        PayType.NON_CLAIMS_BASED_PAYMENTS
      )
    ) {
      history.push(
        `/models/${modelID}/collaboration-area/task-list/payment/non-claims-based-payment`
      );
    } else {
      history.push(
        `/models/${modelID}/collaboration-area/task-list/payment/complexity`
      );
    }
  };

  const initialValues: BeneficiaryCostSharingFormType = {
    __typename: 'PlanPayments',
    id: id ?? '',
    payType: payType ?? [],
    payClaims: payClaims ?? [],
    beneficiaryCostSharingLevelAndHandling:
      beneficiaryCostSharingLevelAndHandling ?? '',
    waiveBeneficiaryCostSharingForAnyServices:
      waiveBeneficiaryCostSharingForAnyServices ?? null,
    waiveBeneficiaryCostSharingServiceSpecification:
      waiveBeneficiaryCostSharingServiceSpecification ?? '',
    waiverOnlyAppliesPartOfPayment: waiverOnlyAppliesPartOfPayment ?? null,
    waiveBeneficiaryCostSharingNote: waiveBeneficiaryCostSharingNote ?? ''
  };

  if ((!loading && error) || (!loading && !data?.modelPlan)) {
    return <NotFoundPartial />;
  }

  return (
    <>
      <MutationErrorModal
        isOpen={mutationError.isModalOpen}
        closeModal={() => mutationError.setIsModalOpen(false)}
        url={mutationError.destinationURL}
      />

      <Breadcrumbs
        items={[
          BreadcrumbItemOptions.HOME,
          BreadcrumbItemOptions.COLLABORATION_AREA,
          BreadcrumbItemOptions.TASK_LIST,
          BreadcrumbItemOptions.PAYMENTS
        ]}
      />
      <PageHeading className="margin-top-4 margin-bottom-2">
        {paymentsMiscT('heading')}
      </PageHeading>

      <p
        className="margin-top-0 margin-bottom-1 font-body-lg"
        data-testid="model-plan-name"
      >
        {miscellaneousT('for')} {modelName}
      </p>

      <p className="margin-bottom-2 font-body-md line-height-sans-4">
        {miscellaneousT('helpText')}
      </p>

      <AskAQuestion modelID={modelID} />

      <Formik
        initialValues={initialValues}
        onSubmit={() => {
          nextPage();
        }}
        enableReinitialize
        innerRef={formikRef}
      >
        {(formikProps: FormikProps<BeneficiaryCostSharingFormType>) => {
          const { handleSubmit, setFieldValue, setErrors, values } =
            formikProps;

          return (
            <>
              <ConfirmLeave />

              <GridContainer className="padding-left-0 padding-right-0">
                <Grid row gap>
                  <Grid desktop={{ col: 6 }}>
                    <Form
                      className="margin-top-6"
                      data-testid="payment-beneficiary-cost-sharing-form"
                      onSubmit={e => {
                        handleSubmit(e);
                      }}
                    >
                      <Fieldset disabled={!!error || loading}>
                        <PageHeading
                          headingLevel="h3"
                          className="margin-bottom-3"
                        >
                          {paymentsMiscT('beneficaryCostSharingQuestions')}
                        </PageHeading>

                        <FieldGroup>
                          <Label
                            htmlFor="beneficiaryCostSharingLevelAndHandling"
                            className="maxw-none"
                          >
                            {paymentsT(
                              'beneficiaryCostSharingLevelAndHandling.label'
                            )}
                          </Label>

                          <Field
                            as={TextAreaField}
                            className="height-15"
                            id="payment-beneficiary-cost-sharing"
                            data-testid="payment-beneficiary-cost-sharing"
                            name="beneficiaryCostSharingLevelAndHandling"
                          />
                        </FieldGroup>

                        <FieldGroup className="margin-top-4">
                          <Label
                            htmlFor="waiveBeneficiaryCostSharingForAnyServices"
                            className="maxw-none"
                          >
                            {paymentsT(
                              'waiveBeneficiaryCostSharingForAnyServices.label'
                            )}
                          </Label>

                          <BooleanRadio
                            field="waiveBeneficiaryCostSharingForAnyServices"
                            id="payment-waive-any-service"
                            value={
                              values.waiveBeneficiaryCostSharingForAnyServices
                            }
                            setFieldValue={setFieldValue}
                            options={
                              waiveBeneficiaryCostSharingForAnyServicesConfig.options
                            }
                            childName="waiveBeneficiaryCostSharingServiceSpecification"
                          >
                            {values.waiveBeneficiaryCostSharingForAnyServices ? (
                              <FieldGroup className="margin-left-4 margin-y-1">
                                <Label
                                  htmlFor="waiveBeneficiaryCostSharingServiceSpecification"
                                  className="text-normal"
                                >
                                  {paymentsT(
                                    'waiveBeneficiaryCostSharingServiceSpecification.label'
                                  )}
                                </Label>

                                <Field
                                  as={TextAreaField}
                                  className="height-15"
                                  id="payment-waive-any-service-specification"
                                  data-testid="payment-waive-any-service-specification"
                                  name="waiveBeneficiaryCostSharingServiceSpecification"
                                />
                              </FieldGroup>
                            ) : (
                              <></>
                            )}
                          </BooleanRadio>

                          {values.waiveBeneficiaryCostSharingForAnyServices && (
                            <FieldGroup className="margin-top-4">
                              <Label
                                htmlFor="waiverOnlyAppliesPartOfPayment"
                                className="maxw-none text-normal"
                              >
                                {paymentsT(
                                  'waiverOnlyAppliesPartOfPayment.label'
                                )}
                              </Label>

                              <p className="text-base margin-y-1">
                                {paymentsT(
                                  'waiverOnlyAppliesPartOfPayment.sublabel'
                                )}
                              </p>

                              <BooleanRadio
                                field="waiverOnlyAppliesPartOfPayment"
                                id="payment-waive-part-of-payment"
                                value={values.waiverOnlyAppliesPartOfPayment}
                                setFieldValue={setFieldValue}
                                options={
                                  waiverOnlyAppliesPartOfPaymentConfig.options
                                }
                              />
                            </FieldGroup>
                          )}
                          <AddNote
                            id="payment-waive-beneficiary-cost-sharing-note"
                            field="waiveBeneficiaryCostSharingNote"
                          />
                        </FieldGroup>

                        <div className="margin-top-6 margin-bottom-3">
                          <Button
                            type="button"
                            className="usa-button usa-button--outline margin-bottom-1"
                            onClick={() => {
                              history.push(
                                `/models/${modelID}/collaboration-area/task-list/payment/anticipating-dependencies`
                              );
                            }}
                          >
                            {miscellaneousT('back')}
                          </Button>

                          <Button type="submit" onClick={() => setErrors({})}>
                            {miscellaneousT('next')}
                          </Button>
                        </div>

                        <Button
                          type="button"
                          className="usa-button usa-button--unstyled"
                          onClick={() =>
                            history.push(
                              `/models/${modelID}/collaboration-area/task-list`
                            )
                          }
                        >
                          <Icon.ArrowBack
                            className="margin-right-1"
                            aria-hidden
                          />

                          {miscellaneousT('saveAndReturn')}
                        </Button>
                      </Fieldset>
                    </Form>
                  </Grid>
                </Grid>
              </GridContainer>
            </>
          );
        }}
      </Formik>

      {data && (
        <PageNumber
          currentPage={renderCurrentPage(
            4,
            payType.includes(PayType.CLAIMS_BASED_PAYMENTS),
            payType.includes(PayType.NON_CLAIMS_BASED_PAYMENTS),
            payClaims.includes(
              ClaimsBasedPayType.REDUCTIONS_TO_BENEFICIARY_COST_SHARING
            )
          )}
          totalPages={renderTotalPages(
            payType.includes(PayType.CLAIMS_BASED_PAYMENTS),
            payType.includes(PayType.NON_CLAIMS_BASED_PAYMENTS),
            payClaims.includes(
              ClaimsBasedPayType.REDUCTIONS_TO_BENEFICIARY_COST_SHARING
            )
          )}
          className="margin-y-6"
        />
      )}
    </>
  );
};
export default BeneficiaryCostSharing;
