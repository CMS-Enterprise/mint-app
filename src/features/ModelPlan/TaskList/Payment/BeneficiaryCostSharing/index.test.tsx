import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import { render, screen, waitFor } from '@testing-library/react';
import {
  ClaimsBasedPayType,
  GetBeneficiaryCostSharingDocument,
  GetBeneficiaryCostSharingQuery,
  PayType
} from 'gql/generated/graphql';

import BeneficiaryCostSharing from './index';

type GetBeneficiaryCostSharingType = GetBeneficiaryCostSharingQuery['modelPlan']['payments'];

const mockData: GetBeneficiaryCostSharingType = {
  __typename: 'PlanPayments',
  id: '123',
  payType: [PayType.CLAIMS_BASED_PAYMENTS],
  payClaims: [ClaimsBasedPayType.OTHER],
  beneficiaryCostSharingLevelAndHandling: 'This is a string',
  waiveBeneficiaryCostSharingForAnyServices: true,
  waiveBeneficiaryCostSharingServiceSpecification: null,
  waiverOnlyAppliesPartOfPayment: null,
  waiveBeneficiaryCostSharingNote: null
};

const paymentsMock = [
  {
    request: {
      query: GetBeneficiaryCostSharingDocument,
      variables: { id: 'ce3405a0-3399-4e3a-88d7-3cfc613d2905' }
    },
    result: {
      data: {
        modelPlan: {
          id: 'ce3405a0-3399-4e3a-88d7-3cfc613d2905',
          modelName: 'My excellent plan that I just initiated',
          payments: mockData
        }
      }
    }
  }
];

describe('Model Plan -- BeneficiaryCostSharing', () => {
  it('renders without errors', async () => {
    render(
      <MemoryRouter
        initialEntries={[
          '/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/collaboration-area/task-list/payment/beneficiary-cost-sharing'
        ]}
      >
        <MockedProvider mocks={paymentsMock} addTypename={false}>
          <Route path="/models/:modelID/collaboration-area/task-list/payment/beneficiary-cost-sharing">
            <BeneficiaryCostSharing />
          </Route>
        </MockedProvider>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(
        screen.getByTestId('payment-beneficiary-cost-sharing-form')
      ).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(
        screen.getByTestId('payment-beneficiary-cost-sharing')
      ).toHaveValue('This is a string');
    });

    await waitFor(() => {
      expect(
        screen.getByTestId('payment-waive-any-service-specification')
      ).toBeInTheDocument();
    });
  });

  it('matches snapshot', async () => {
    const { asFragment } = render(
      <MemoryRouter
        initialEntries={[
          '/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/collaboration-area/task-list/payment/beneficiary-cost-sharing'
        ]}
      >
        <MockedProvider mocks={paymentsMock} addTypename={false}>
          <Route path="/models/:modelID/collaboration-area/task-list/payment/beneficiary-cost-sharing">
            <BeneficiaryCostSharing />
          </Route>
        </MockedProvider>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(
        screen.getByTestId('payment-beneficiary-cost-sharing')
      ).toHaveValue('This is a string');
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
