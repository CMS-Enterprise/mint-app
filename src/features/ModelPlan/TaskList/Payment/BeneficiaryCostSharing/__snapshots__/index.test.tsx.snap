// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Model Plan -- BeneficiaryCostSharing > matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="mint-main-content"
    data-testid="model-payment"
    id="main-content"
    role="main"
    tabindex="-1"
  >
    <div
      class="grid-container"
      data-testid="gridContainer"
    >
      <nav
        aria-label="Breadcrumbs"
        class="usa-breadcrumb usa-breadcrumb--wrap"
      >
        <ol
          class="usa-breadcrumb__list"
        >
          <li
            class="usa-breadcrumb__list-item"
          >
            <a
              class="usa-link usa-breadcrumb__link"
              data-discover="true"
              href="/"
            >
              <span>
                Home
              </span>
            </a>
          </li>
          <li
            class="usa-breadcrumb__list-item"
          >
            <a
              class="usa-link usa-breadcrumb__link"
              data-discover="true"
              href="/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/collaboration-area"
            >
              <span>
                Model collaboration area
              </span>
            </a>
          </li>
          <li
            class="usa-breadcrumb__list-item"
          >
            <a
              class="usa-link usa-breadcrumb__link"
              data-discover="true"
              href="/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/collaboration-area/task-list"
            >
              <span>
                Model Plan task list
              </span>
            </a>
          </li>
          <li
            aria-current="page"
            class="usa-breadcrumb__list-item usa-current"
          >
            <span>
              Payment
            </span>
          </li>
        </ol>
      </nav>
      <h1
        aria-live="polite"
        class="mint-h1 margin-top-4 margin-bottom-2"
        tabindex="-1"
      >
        Payment
      </h1>
      <p
        class="margin-top-0 margin-bottom-1 font-body-lg"
        data-testid="model-plan-name"
      >
        for My excellent plan that I just initiated
      </p>
      <p
        class="margin-bottom-2 font-body-md line-height-sans-4"
      >
        If there's a question or field that is not applicable to your model or you don't currently know the answer, you may leave it blank. If you need help, ask a question using the link below.
      </p>
      <div>
        <div
          class="padding-2 bg-primary-lighter display-flex flex-wrap flex-justify flex-align-center"
          style="gap: 1rem;"
        >
          <div
            class="display-flex"
            data-testid="ask-a-question"
          >
            <svg
              aria-label="announcement"
              class="usa-icon text-primary margin-right-1"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-7 9h-2V5h2v6zm0 4h-2v-2h2v2z"
              />
            </svg>
            <button
              class="usa-button usa-button--unstyled"
              data-testid="ask-a-question-button"
              type="button"
            >
              Start a discussion
            </button>
          </div>
        </div>
      </div>
      <div
        class="grid-container padding-left-0 padding-right-0"
        data-testid="gridContainer"
      >
        <div
          class="grid-row grid-gap"
          data-testid="grid"
        >
          <div
            class="desktop:grid-col-6"
            data-testid="grid"
          >
            <form
              class="margin-top-6"
              data-testid="payment-beneficiary-cost-sharing-form"
            >
              <fieldset
                class="usa-fieldset"
                data-testid="fieldset"
              >
                <h3
                  aria-live="polite"
                  class="mint-h1 margin-bottom-3"
                  tabindex="-1"
                >
                  Beneficiary cost-sharing questions
                </h3>
                <div
                  class="usa-form-group"
                >
                  <label
                    class="usa-label maxw-none"
                    data-testid="label"
                    for="beneficiaryCostSharingLevelAndHandling"
                  >
                    What is the intended level of beneficiary cost-sharing and how will it be handled?
                  </label>
                  <textarea
                    aria-labelledby="label-payment-beneficiary-cost-sharing hint-payment-beneficiary-cost-sharing"
                    class="height-15 usa-textarea"
                    data-testid="payment-beneficiary-cost-sharing"
                    id="payment-beneficiary-cost-sharing"
                    name="beneficiaryCostSharingLevelAndHandling"
                  >
                    This is a string
                  </textarea>
                </div>
                <div
                  class="usa-form-group margin-top-4"
                >
                  <label
                    class="usa-label maxw-none"
                    data-testid="label"
                    for="waiveBeneficiaryCostSharingForAnyServices"
                  >
                    Will you waive beneficiary cost-sharing (coinsurance and/or deductible) for any services in the model?
                  </label>
                  <fieldset
                    class="usa-fieldset"
                    data-testid="fieldset"
                  >
                    <div
                      class="usa-radio"
                      data-testid="radio"
                    >
                      <input
                        class="usa-radio__input"
                        data-testid="payment-waive-any-service-true"
                        id="payment-waive-any-service-true"
                        name="waiveBeneficiaryCostSharingForAnyServices"
                        type="radio"
                        value="TRUE"
                      />
                      <label
                        class="usa-radio__label"
                        for="payment-waive-any-service-true"
                      >
                        Yes
                      </label>
                    </div>
                    <div
                      class="usa-form-group margin-left-4 margin-y-1"
                    >
                      <label
                        class="usa-label text-normal"
                        data-testid="label"
                        for="waiveBeneficiaryCostSharingServiceSpecification"
                      >
                        Please specify which services.
                      </label>
                      <textarea
                        aria-labelledby="label-payment-waive-any-service-specification hint-payment-waive-any-service-specification"
                        class="height-15 usa-textarea"
                        data-testid="payment-waive-any-service-specification"
                        id="payment-waive-any-service-specification"
                        name="waiveBeneficiaryCostSharingServiceSpecification"
                      />
                    </div>
                    <div
                      class="usa-radio"
                      data-testid="radio"
                    >
                      <input
                        class="usa-radio__input"
                        data-testid="payment-waive-any-service-false"
                        id="payment-waive-any-service-false"
                        name="waiveBeneficiaryCostSharingForAnyServices"
                        type="radio"
                        value="FALSE"
                      />
                      <label
                        class="usa-radio__label"
                        for="payment-waive-any-service-false"
                      >
                        No
                      </label>
                    </div>
                  </fieldset>
                  <div
                    class="usa-form-group margin-top-4"
                  >
                    <label
                      class="usa-label maxw-none text-normal"
                      data-testid="label"
                      for="waiverOnlyAppliesPartOfPayment"
                    >
                      Would the waiver only apply for part of the payment?
                    </label>
                    <p
                      class="text-base margin-y-1"
                    >
                      For example, we would assess cost-sharing for the service itself, but not for an add-on payment
                    </p>
                    <fieldset
                      class="usa-fieldset"
                      data-testid="fieldset"
                    >
                      <div
                        class="usa-radio"
                        data-testid="radio"
                      >
                        <input
                          class="usa-radio__input"
                          data-testid="payment-waive-part-of-payment-true"
                          id="payment-waive-part-of-payment-true"
                          name="waiverOnlyAppliesPartOfPayment"
                          type="radio"
                          value="TRUE"
                        />
                        <label
                          class="usa-radio__label"
                          for="payment-waive-part-of-payment-true"
                        >
                          Yes
                        </label>
                      </div>
                      <div
                        class="usa-radio"
                        data-testid="radio"
                      >
                        <input
                          class="usa-radio__input"
                          data-testid="payment-waive-part-of-payment-false"
                          id="payment-waive-part-of-payment-false"
                          name="waiverOnlyAppliesPartOfPayment"
                          type="radio"
                          value="FALSE"
                        />
                        <label
                          class="usa-radio__label"
                          for="payment-waive-part-of-payment-false"
                        >
                          No
                        </label>
                      </div>
                    </fieldset>
                  </div>
                  <div
                    class="margin-top-3 margin-bottom-6"
                  >
                    <button
                      class="usa-button usa-button usa-button--unstyled"
                      data-testid="payment-waive-beneficiary-cost-sharing-note-add-note-toggle"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        aria-label="add"
                        class="usa-icon margin-right-1"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                        />
                      </svg>
                      Add an additional note
                    </button>
                  </div>
                </div>
                <div
                  class="margin-top-6 margin-bottom-3"
                >
                  <button
                    class="usa-button usa-button usa-button--outline margin-bottom-1"
                    data-testid="button"
                    type="button"
                  >
                    Back
                  </button>
                  <button
                    class="usa-button"
                    data-testid="button"
                    type="submit"
                  >
                    Next
                  </button>
                </div>
                <button
                  class="usa-button usa-button usa-button--unstyled"
                  data-testid="button"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    aria-label="back"
                    class="usa-icon margin-right-1"
                    focusable="false"
                    height="1em"
                    role="img"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                    />
                  </svg>
                  Save and return to task list
                </button>
              </fieldset>
            </form>
          </div>
        </div>
      </div>
      <div
        class="mint-page-number margin-y-6"
        data-testid="page-num"
      >
        <span
          class="mint-page-number__page-num"
        >
          Page 3 of 5
        </span>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
